llvm/include/llvm/CodeGen/ISDOpcodes.h

```diff
@@ -804,6 +804,9 @@ enum NodeType {
 
   /// TRUNCATE - Completely drop the high bits.
   TRUNCATE,
+  /// TRUNCATE_[SU]SAT - Truncate for saturated operand
+  TRUNCATE_SSAT,
```

Comment:
Is TRUNCATE_SSAT signed input to signed result?
Is TRUNCATE_USAT unsigned input to unsigned result?

I ask because X86's packuswb instructon does signed input to unsigned result so its worth being clearing here.

Reply:
*This is my opinion and needs confirmation*. 
It is true that SSAT/USAT is signed, but 'truncate_[us]sat' means that the range of values for 'truncate' is in the range of values that don't have to care about the sign bit.

We also don't support unsigned type variables, right? Therefore, the 'truncate_sat' instruction doesn't care about the type of target value, just convert it to return type and return it.

Reply:
Variables don't have signed or unsigned type but operations can interpret their input as signed or unsigned.

I believe as you have defined it TRUNCATE_USAT will interpret the input as an unsigned value and produce an unsigned result in the destination type.

The x86 packuswb interprets the input as signed 16 bits and produces an unsigned 8 bit result. If the input is negative it will return 0. That is a different operation than either operation defined here. It's equivalent to smax with 0 followed by truncate_usat. I could imagine having that operation as a single node. Not suggesting for this patch.

I just want the semantics of the new opcodes documented.

Reply:
AArch64 has a SQXTUN node too: https://docsmirror.github.io/A64/2023-09/sqxtun_advsimd.html

Reply:
So should we have 3 truncsat nodes? I mentioned this on #85903 but hadn't realised other targets had something similar to PACKUS
```
TRUNCATE_SSAT_S,  // saturate signed input to signed result - truncate(smax(smin(x)))
TRUNCATE_SSAT_U,  // saturate signed input to unsigned result - truncate(umax(smin(x,0)))
TRUNCATE_USAT_U,  // saturate unsigned input to unsigned result - truncate(umin(x))
```

Reply:
@RKSimon should add that ISD? is it not necessary for unsigned input to signed input?

Reply:
If we have targets that support for TRUNCATE_USAT_S then by all means support it.